### PR TITLE
[BREAKING] Underscores in numeric literals

### LINKED ISSUE #1677: Underscores in number literals
It would be nice to support underscores in number literals to have easily readable numbers: `1_000_000`.

Question:
- should it allow multiple underscores next to each other `1__000_____000`?
- should it allow underscores at arbitrary places `1_0_00_000_0_000`?

### PR DESCRIPTION
**Pull request by christianparpart**
Closes #1677

Follows https://www.python.org/dev/peps/pep-0515/

This is a recreation of the PR #3120, that got accidentally closed and couldn't be recovered (easily). My apologies, @Balajiganapathi. I however retained your original commits and will now attempt finishing them.

### Notes

This PR finalizes the original PR #3120. Changes include:
* Shifting validation of underscores as far back as possible (from Lexical analysis to SyntaxChecker) to provide better error diagnostics.
* Adds tests to `tests/libsolidity/syntaxChecker/parsing/`.
* We agreed to keep rules as similar as possible to recent Python, i.e. minimal validation is done, no thus, grouping width is forced neither on decimal nor on hexadecimal.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -46,3 +46,5 @@ browse.VC.db
 CMakeLists.txt.user
 /CMakeSettings.json
 /.vs
+/.cproject
+/.project
```
Comment (ekpyron) [inline .gitignore:50]: Not really part of this PR, but probably fine to just leave it in here.

File: docs/types.rst [hunk 2]
```diff
@@ -276,6 +276,9 @@ one side.  Examples include ``1.``, ``.1`` and ``1.3``.
 Scientific notation is also supported, where the base can have fractions, while the exponent cannot.
 Examples include ``2e10``, ``-2e10``, ``2e-10``, ``2.5e1``.
 
+Underscores can be used to separate the digits of a numeric literal to aid readability.
+For example, ``123_000``, ``0x2eff_abde``, ``1233_e348_9a`` are all valid. Underscores are only allowed between two digits. For hex literals, underscores are only allowed to separate groups of 4 hex digits.
```
Comment (ekpyron) [inline docs/types.rst]: I first misread this and thought that there always have to be two digits on each side of the underscore. Maybe ``Literals may not start or end in an underscore.`` (or something better) instead?

File: libsolidity/analysis/SyntaxChecker.cpp [hunk 3]
```diff
@@ -183,6 +186,58 @@ bool SyntaxChecker::visit(Throw const& _throwStatement)
...
+	{
+		m_errorReporter.syntaxError(_literal.location(), "Invalid use of underscores in number literal. Only one consecutive underscores between digits allowed.");
+		return true;
+	}
+
+	if (!_literal.isHexNumber())
+		// No special validation on grouped decimals yet.
+		return true;
+
+	// We have a hex literal. Do underscore validation
```
Comment (ekpyron) [inline libsolidity/analysis/SyntaxChecker.cpp]: Do we need this comment at all? It's a bit misleading here, since underscore validation only starts a few lines below, so it should probably at least be moved there.

File: libsolidity/analysis/SyntaxChecker.cpp [hunk 4]
```diff
@@ -183,6 +186,58 @@ bool SyntaxChecker::visit(Throw const& _throwStatement)
...
+	boost::split(parts, value, boost::is_any_of("_"));
+
+	if (parts.size() == 1) // no underscores
+		return true;
+
+	if (parts.front().size() > 4)
+		m_errorReporter.syntaxError(_literal.location(), "Invalid use of underscores in hex literal. The first part " + parts[0] + " is too large. Group up by thousands.");
+
+	// Everything except first and last part must be 4 chars in length
+	for (size_t i = 1; i + 1 < parts.size(); ++i)
```
Comment (ekpyron) [inline libsolidity/analysis/SyntaxChecker.cpp]: Maybe ``i < parts.size() - 1`` would be easier to read.

File: test/libsolidity/SolidityScanner.cpp [hunk 5]
```diff
@@ -155,6 +155,50 @@ BOOST_AUTO_TEST_CASE(trailing_dot)
 	BOOST_CHECK_EQUAL(scanner.next(), Token::EOS);
 }
 
+BOOST_AUTO_TEST_CASE(leading_underscore_decimal_is_identifier)
+{
+	Scanner scanner(CharStream("var x = _1.2;"));
```
Comment (ekpyron) [inline test/libsolidity/SolidityScanner.cpp]: Do we want to use ``var`` here? It's merely the scanner alright, but still. Although we'll probably keep ``var`` as keywords for the suggestions, so maybe it's fine. Same in the tests below.

File: test/libsolidity/syntaxTests/parsing/lexer_numbers_with_underscores_decimal.sol [hunk 6]
```diff
@@ -0,0 +1,20 @@
+contract C {
+  function f() public pure {
+    // good
```
Comment (ekpyron) [inline test/libsolidity/syntaxTests/parsing/lexer_numbers_with_underscores_decimal.sol]: Maybe better to split the tests into a succeeding and a failing part. Same for the other tests.

File: libsolidity/analysis/SyntaxChecker.cpp (hunk 4 above)
Comment (christianparpart) [reply libsolidity/analysis/SyntaxChecker.cpp]: I agree.

File: test/libsolidity/SolidityScanner.cpp (hunk 5 above)
Comment (christianparpart) [reply test/libsolidity/SolidityScanner.cpp]: Certainly not. This is a left-over from the original PR (original author). I didn't clear all of them yet. Ideally those scanner tests should only tokenize the minimum required tokens.

File: test/libsolidity/syntaxTests/parsing/lexer_numbers_with_underscores_decimal.sol (hunk 6 above)
Comment (christianparpart) [reply test/libsolidity/syntaxTests/parsing/lexer_numbers_with_underscores_decimal.sol]: Sadly not sure what benefits this would give us. But I saw other tests doing that too, yes.

File: .gitignore (hunk 1 above)
Comment (christianparpart) [reply .gitignore:50]: This is part of the original PR. I agree however. On the other hand, if we don't include them now, they'll never be included, because such a tiny PR would be very far at the end of the PRs to be reviewed. :)

File: docs/types.rst (hunk 2 above)
Comment (ekpyron) [reply docs/types.rst]: Hm... my suggestion doesn't exclude ``__`` (two underscores)... not sure - maybe it's fine as it is, maybe you find something better.

Comment (christianparpart) [reply docs/types.rst]: I am also not yet sure whether I like the idea of having such relaxed grouping rules for decimals. I'll have a look over the text. Thanks.

File: test/libsolidity/syntaxTests/parsing/lexer_numbers_with_underscores_decimal.sol (hunk 6 above)
Comment (ekpyron) [reply test/libsolidity/syntaxTests/parsing/lexer_numbers_with_underscores_decimal.sol]: It's easier to review and more apparent that everything still works as intended, if something changes. It's difficult to verify the source locations in git diffs, so it may go unnoticed, if a previously succeeding test fails and a previously failing test succeeds (even though that may be unlikely in practice). Originally ``isoltest`` used line numbers, not source location, maybe we should have stayed with that or at least kept them :-D... but it would be a major hassle to change it now :-).

File: docs/types.rst (hunk 2 above)
Comment (ekpyron) [reply docs/types.rst]: I think the relaxed rules are fine - stricter rules are probably more annoying than helpful I guess.

File: .gitignore (hunk 1 above)
Comment (axic) [reply .gitignore:50]: What IDE uses these?

File: test/libsolidity/syntaxTests/parsing/lexer_numbers_with_underscores_decimal.sol (hunk 6 above)
Comment (axic) [reply test/libsolidity/syntaxTests/parsing/lexer_numbers_with_underscores_decimal.sol]: Yes please split them. The succeeding ones should never be updated in PRs, unless we make a breaking change, so it will be apparent if anything broke. The failing ones may be updated more frequently, because of better error messages, etc.

File: .gitignore (hunk 1 above)
Comment (christianparpart) [reply .gitignore:50]: Eclipse CDT and maybe the latter by IntelliJ IDEA (Never used either of them, but google helped me out) :-)

Comment (Balajiganapathi) [reply .gitignore:50]: Yes, I had used Eclipse CDT.

File: libsolidity/analysis/SyntaxChecker.cpp [hunk 7]
```diff
@@ -188,40 +188,63 @@ bool SyntaxChecker::visit(Throw const& _throwStatement)
 
 bool SyntaxChecker::visit(Literal const& _literal)
 {
+	if (_literal.token() != Token::Number)
+		return true;
+
+	// Generic checks no matter what base this number literal is of:
+	if (_literal.value().back() == '_')
```
Comment (axic) [inline libsolidity/analysis/SyntaxChecker.cpp]: Why not create a local variable for this since it is referenced so many times?

File: test/libsolidity/syntaxTests/parsing/lexer_numbers_with_underscores_decimal.sol (hunk 6 above)
Comment (christianparpart) [reply test/libsolidity/syntaxTests/parsing/lexer_numbers_with_underscores_decimal.sol]: :+1:

File: docs/types.rst [hunk 8]
```diff
@@ -284,6 +284,12 @@ one side.  Examples include ``1.``, ``.1`` and ``1.3``.
 Scientific notation is also supported, where the base can have fractions, while the exponent cannot.
 Examples include ``2e10``, ``-2e10``, ``2e-10``, ``2.5e1``.
 
+Underscores can be used to separate the digits of a numeric literal to aid readability.
+For example, decimal ``123_000``, hexadecimal ``0x2eff_abde``, scientific decimal notation ``1233_e348_9a`` are all valid.
+Underscores are only allowed between two digits and only one consecutive underscore is allowed.
+For hex literals, underscores are only allowed to separate groups of 4 hex digits.
+Also, no underscore may occur more than once to delimit
```
Comment (chriseth) [inline docs/types.rst]: Partial sentence.

File: libsolidity/parsing/Scanner.cpp [hunk 9]
```diff
@@ -768,6 +779,15 @@ Token::Value Scanner::scanNumber(char _charSeen)
 			scanDecimalDigits();  // optional
 			if (m_char == '.')
 			{
+				if (m_source.get(1) == '_')
```
Comment (chriseth) [inline libsolidity/parsing/Scanner.cpp]: Can this go out of bounds?

File: test/libsolidity/SolidityScanner.cpp [hunk 10]
```diff
@@ -155,6 +155,44 @@ BOOST_AUTO_TEST_CASE(trailing_dot)
...
+}
+
+BOOST_AUTO_TEST_CASE(leading_underscore_hex_illegal)
+{
+	Scanner scanner(CharStream("0x_abc"));
+	BOOST_CHECK_EQUAL(scanner.currentToken(), Token::Illegal);
+	BOOST_CHECK_EQUAL(scanner.next(), Token::Identifier);
+	BOOST_CHECK_EQUAL(scanner.next(), Token::EOS);
+}
+
```
Comment (chriseth) [inline test/libsolidity/SolidityScanner.cpp:203]: Please add a test with `_` at end of input.

File: docs/types.rst [hunk 11]
```diff
@@ -284,6 +284,11 @@ one side.  Examples include ``1.``, ``.1`` and ``1.3``.
 Scientific notation is also supported, where the base can have fractions, while the exponent cannot.
 Examples include ``2e10``, ``-2e10``, ``2e-10``, ``2.5e1``.
 
+Underscores can be used to separate the digits of a numeric literal to aid readability.
+For example, decimal ``123_000``, hexadecimal ``0x2eff_abde``, scientific decimal notation ``1233_e348_9a`` are all valid.
+Underscores are only allowed between two digits and only one consecutive underscore is allowed.
+For hex literals, underscores are only allowed to separate groups of 4 hex digits.
```
Comment (chriseth) [inline docs/types.rst]: 4 or more?

File: libsolidity/analysis/SyntaxChecker.cpp [hunk 12]
```diff
@@ -183,6 +186,71 @@ bool SyntaxChecker::visit(Throw const& _throwStatement)
...
 
+bool SyntaxChecker::visit(Literal const& _literal)
+{
+	if (_literal.token() != Token::Number)
+		return true;
+
+	ASTString value = _literal.value();
+
+	// Generic checks no matter what base this number literal is of:
+	if (value.back() == '_')
```
Comment (chriseth) [inline libsolidity/analysis/SyntaxChecker.cpp]: I know it might be silly, but please assert `!value.empty()`

File: libsolidity/analysis/SyntaxChecker.cpp [hunk 13]
```diff
@@ -183,6 +186,71 @@ bool SyntaxChecker::visit(Throw const& _throwStatement)
...
+			m_errorReporter.syntaxError(_literal.location(), "Invalid use of underscores in number literal. No underscores in front of the fraction part allowed.");
+
+		if (value.find("_e") != ASTString::npos)
+			m_errorReporter.syntaxError(_literal.location(), "Invalid use of underscores in number literal. No underscore in front of exponent allowed.");
+
+		// No special validation on grouped decimals yet.
+		return true;
+	}
+
+	// We have a hex literal. Do underscore validation
```
Comment (chriseth) [inline libsolidity/analysis/SyntaxChecker.cpp]: Perhaps split this into its own function?

File: libsolidity/analysis/SyntaxChecker.cpp [hunk 14]
```diff
@@ -183,6 +186,71 @@ bool SyntaxChecker::visit(Throw const& _throwStatement)
...
+	solAssert(value.substr(0, 2) == "0x", "");
+
+	value = value.substr(2); // Skip the 0x
+	vector<ASTString> parts;
+	boost::split(parts, value, boost::is_any_of("_"));
```
Comment (chriseth) [inline libsolidity/analysis/SyntaxChecker.cpp]: While I usually don't care much about efficiency, I wonder whether this part might be better done with iterators or counters.

**Review by chriseth — DISMISSED** [review]

File: libsolidity/parsing/Scanner.cpp (hunk 9 above)
Comment (christianparpart) [reply libsolidity/parsing/Scanner.cpp]: Probably in special crafted source codes (that are then invalid anyways). I've been adding a guard now.

File: test/libsolidity/SolidityScanner.cpp (hunk 10 above)
Comment (christianparpart) [reply test/libsolidity/SolidityScanner.cpp:203]: done.

File: docs/types.rst [hunk 15]
```diff
@@ -284,6 +284,11 @@ one side.  Examples include ``1.``, ``.1`` and ``1.3``.
 Scientific notation is also supported, where the base can have fractions, while the exponent cannot.
 Examples include ``2e10``, ``-2e10``, ``2e-10``, ``2.5e1``.
 
+Underscores can be used to separate the digits of a numeric literal to aid readability.
+For example, decimal ``123_000``, hexadecimal ``0x2eff_abde``, scientific decimal notation ``1233_e348_9a`` are all valid.
```
Comment (chriseth) [inline docs/types.rst]: Is `1233_e348` really valid (underscore right in front of the `e`)? And also the `a` at the end shouldn't really be there, should it?

File: libsolidity/analysis/SyntaxChecker.cpp [hunk 16]
```diff
@@ -183,6 +186,42 @@ bool SyntaxChecker::visit(Throw const& _throwStatement)
...
+
+	if (value.find("__") != ASTString::npos)
+	{
+		m_errorReporter.syntaxError(_literal.location(), "Invalid use of underscores in number literal. Only one consecutive underscores between digits allowed.");
+		return true;
+	}
+
+	if (!_literal.isHexNumber()) // decimal literal
+	{
+		if (value.find("_.") != ASTString::npos)
```
Comment (chriseth) [inline libsolidity/analysis/SyntaxChecker.cpp]: Why is `._` allowed? Because it is allowed in python?

**Review by chriseth — CHANGES_REQUESTED** [review]

File: docs/types.rst (hunk 15 above)
Comment (christianparpart) [reply docs/types.rst]: No. We do not allow underscores at the edges of a number sequence. This documentation was wrong. Fixed. Thanks!

**Review by chriseth — APPROVED** [review]
